File: app/api/routes/stats.py
import logging

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.dao import stats_dao
from app.db.async_session import get_db
from app.schemas.stats import (
    ListeningTrendItem,
    ListeningTrendsResponse,
    TopArtistItem,
    TopArtistsResponse,
    TopTrackItem,
    TopTracksResponse,
)

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/stats", tags=["stats"])


@router.get("/top-tracks", response_model=TopTracksResponse)
async def get_top_tracks(limit: int = 10, db: AsyncSession = Depends(get_db)):
    rows = await stats_dao.get_top_tracks(db, limit)
    return TopTracksResponse(tracks=[TopTrackItem(**row) for row in rows])


@router.get("/top-artists", response_model=TopArtistsResponse)
async def get_top_artists(limit: int = 10, db: AsyncSession = Depends(get_db)):
    rows = await stats_dao.get_top_artists(db, limit)
    return TopArtistsResponse(artists=[TopArtistItem(**row) for row in rows])


@router.get("/listening-trends", response_model=ListeningTrendsResponse)
async def get_listening_trends(db: AsyncSession = Depends(get_db)):
    rows = await stats_dao.get_listening_trends(db)
    return ListeningTrendsResponse(trends=[ListeningTrendItem(**row) for row in rows])

File: frontend/src/api/stats.ts
import { apiGet } from './client';
import type { ListeningTrendItem, TopArtistItem, TopTrackItem } from './types';

export function getTopTracks(limit = 10): Promise<{ tracks: TopTrackItem[] }> {
  return apiGet(`/stats/top-tracks?limit=${limit}`);
}

export function getTopArtists(limit = 10): Promise<{ artists: TopArtistItem[] }> {
  return apiGet(`/stats/top-artists?limit=${limit}`);
}

export function getListeningTrends(): Promise<{ trends: ListeningTrendItem[] }> {
  return apiGet('/stats/listening-trends');
}
